src/Xrefcheck/Core.hs

```diff
@@ -71,66 +70,58 @@ instance Given ColorMode => Buildable Position where
 
 -- | Full info about a reference.
 data Reference = Reference
-  { rName   :: Text
+  { rName :: Text
     -- ^ Text displayed as reference.
-  , rLink   :: Text
-    -- ^ File or site reference points to.
-  , rAnchor :: Maybe Text
-    -- ^ Section or custom anchor tag.
-  , rPos    :: Position
+  , rPos :: Position
     -- ^ Position in source file.
-  , rInfo   :: ReferenceInfo
-    -- ^ More info about the link.
+  , rInfo :: ReferenceInfo
+    -- ^ More info about the reference.
   } deriving stock (Show, Generic)
 
 -- | Info about the reference.
 data ReferenceInfo
-  = RIExternal
-    -- ^ Reference to a file at outer site, e.g @[d](http://www.google.com/doodles)@
-  | RIOtherProtocol
-    -- ^ Entry not to be processed, e.g. @mailto:e-mail@
-  | RIFileLocal
-    -- ^ Reference to this file, e.g. @[a](#header)@
-  | RIFileAbsolute
-    -- ^ Reference to a file absolute to the root, e.g. @[c](/folder/file#header)@
-  | RIFileRelative
-    -- ^ Reference to a file relative to given one, e.g. @[b](folder/file#header)@
+  = RIExternal ExternalLink
+  | RIFile ReferenceInfoFile
+  deriving stock (Show, Generic)
+
+data ReferenceInfoFile = ReferenceInfoFile
+  { rifAnchor :: Maybe Text
+    -- ^ Section or custom anchor tag.
+  , rifLink :: FileLink
+    -- ^ More info about the link.
+  } deriving stock (Show, Generic)
+
+data ExternalLink
+  = ELUrl Text
+    -- ^ Reference to a file at outer site, e.g @[d](http://www.google.com/doodles)@.
+  | ELOther Text
+    -- ^ Entry not to be processed, e.g. @mailto:e-mail@.
+  deriving stock (Show, Generic)
+
+data FileLink
+  = FLAbsolute Text
+    -- ^ Reference to a file relative to the repository root.
```

Comment:
Note that links to directories are also valid, so this naming may be misleading.

I cannot imagine a better naming myself, but at least a haddock description seems mandatory here.